flink-runtime/src/main/java/org/apache/flink/streaming/api/operators/InternalTimerServiceAsyncImpl.java

```diff
@@ -125,13 +137,14 @@ protected void foreachTimer(
                 "Batch operation is not supported when using async state.");
     }
 
-    private void maintainContextAndProcess(
+    private StateFuture<Void> maintainContextAndProcess(
             InternalTimer<K, N> timer, ThrowingRunnable<Exception> runnable) {
         RecordContext<K> recordCtx = asyncExecutionController.buildContext(null, timer.getKey());
         recordCtx.retain();
         asyncExecutionController.setCurrentContext(recordCtx);
         keyContext.setCurrentKey(timer.getKey());
-        asyncExecutionController.syncPointRequestWithCallback(runnable);
+        StateFuture<Void> future = asyncExecutionController.syncPointRequestWithCallback(runnable);
         recordCtx.release();
+        return future;
```

Comment:
NIT: 
I am curious if there a reason to release the context before returning. If not could we have could we change to `return asyncExecutionController.syncPointRequestWithCallback(runnable);`